### PR TITLE
Fix/remove tests for non-sampled transactions

### PR DESCRIPTION
**Pull request by eyalkoren**

### CODE REVIEW

**Review by eyalkoren — COMMENTED** [review]
Besides the comments, please make sure we DO maintain tests for non-sampled transactions like:
- context propagation works for non-sampled transactions
- at least one central place to verify that we actually don't send non-sampled transactions

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java [hunk 1]
```diff
@@ -340,10 +340,10 @@ public boolean supportsLogsEndpoint() {
     }
 
     public boolean supportsKeepingUnsampledTransaction() {
-        // supportsKeepingUnsampledTransaction is called from application threads
-        // return true instead of blocking the thread
+        // Method is called from application threads thus we have to return fast to avoid blocking application threads
+        // When server version is not known we assume that it's a recent one and discard unsampled transactions.
         if (apmServerVersion != null && !apmServerVersion.isDone()) {
-            return true;
+            return false;
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java:346]: I think paying this overhead is much preferable over potential loss of functionality.
++

While the agent hasn't determined the version of the server, it has to assume that the server may need upsampled transactions.

[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java]: Worth making sure other agents are aware of this case as well.
It is not only about startup time, but in general for unknown APM Server version. 
In one of the 7.x versions there was a change in the healthcheck response format, causing the agent to fail parsing the APM Server version. We should always use backwards compatible defaults where possible.

**Comment by SylvainJuge** [comment]
> Besides the comments, please make sure we DO maintain tests for non-sampled transactions like:
> 
> * context propagation works for non-sampled transactions
> * at least one central place to verify that we actually don't send non-sampled transactions

Those are already covered by those tests, thus it should be fine:
- `co.elastic.apm.agent.impl.transaction.TraceContextTest#testPropagateTransactionIdForUnsampledSpan`
- `co.elastic.apm.agent.impl.transaction.TraceContextTest#testPropagateSpanIdForSampledSpan`
- `co.elastic.apm.agent.impl.ElasticApmTracerTest#testSamplingNone_sendUnsampled`
- `co.elastic.apm.agent.impl.ElasticApmTracerTest#testSamplingNone_dropUnsampled`

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (SylvainJuge) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java]: Adding this to the spec for unsampled transactions: https://github.com/elastic/apm/pull/639

File: apm-opentracing/src/test/java/co/elastic/apm/opentracing/OpenTracingBridgeTest.java [hunk 2]
```diff
@@ -487,9 +494,16 @@ void testManualSampling() {
             span.finish();
         }
         transaction.finish();
-        assertThat(reporter.getTransactions()).hasSize(1);
-        assertThat(reporter.getFirstTransaction().getContext().getLabel("foo")).isNull();
-        assertThat(reporter.getSpans()).isEmpty();
+
+        if(serverSupportsUnsampled){
```
Comment (felixbarny) [inline apm-opentracing/src/test/java/co/elastic/apm/opentracing/OpenTracingBridgeTest.java:498]:
```suggestion
        if (serverSupportsUnsampled) {
```

File: apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java [hunk 3]
```diff
@@ -592,21 +590,34 @@ public void awaitUntilAsserted(ThrowingRunnable runnable) {
...
-        if (thrown != null) {
-            throw new RuntimeException(String.format("Condition not fulfilled within %d ms", timeoutMs), thrown);
+        throw new RuntimeException(String.format("Condition not fulfilled within %d ms", timeoutMs), thrown);
+    }
+
+    /**
+     * @param timeoutMs timeout of the condition
+     * @param runnable  a runnable that throws an exception when the condition is not met
+     */
+    public void awaitUntilTimeout(long timeoutMs, ThrowingRunnable runnable) {
```
Comment (felixbarny) [inline apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java:613]: Isn't that effectively the same as
```
Thread.sleep(timeoutMs);
runnable.run();
```

**Review by felixbarny — APPROVED** [review]

File: apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java (hunk 3 above)
Comment (SylvainJuge) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java:613]: That's quite close, with the minor difference that it's failing faster when the assertion throws.

Comment (felixbarny) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java:613]: ah, right

**Pull request merged**